src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -0,0 +1,2966 @@
+                // (11,17): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         y = [2, null]; // 3
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(11, 17));
+        }
+
+        [Fact]
+        public void Nullable_03()
+        {
+            string source = """
+                #nullable enable
+                using System.Collections;
+                struct S<T> : IEnumerable
+                {
+                    public void Add(T t) { }
+                    public T this[int index] => default!;
+                    IEnumerator IEnumerable.GetEnumerator() => default!;
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        S<object?> x = [1];
+                        x[0].ToString(); // 1
+                        S<object> y = [null]; // 2
+                        y[0].ToString();
+                        y = [2, null]; // 3
+                        y[1].ToString();
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (14,9): warning CS8602: Dereference of a possibly null reference.
+                //         x[0].ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x[0]").WithLocation(14, 9),
+                // (15,24): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         S<object> y = [null]; // 2
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(15, 24),
+                // (17,17): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         y = [2, null]; // 3
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(17, 17));
+        }
+
+        [Fact]
+        public void Nullable_04()
+        {
+            string source = """
+                #nullable enable
+                using System.Collections;
+                struct S<T> : IEnumerable
+                {
+                    public void Add(T t) { }
+                    public T this[int index] => default!;
+                    IEnumerator IEnumerable.GetEnumerator() => default!;
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        S<object>? x = [];
+                        x = [];
+                        S<object>? y = [1];
+                        y = [2];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (13,24): error CS9500: Cannot initialize type 'S<object>?' with a collection literal because the type is not constructible.
```

Comment:
I didn't understand why this case was erroneous.

Reply:
The target type is `Nullable<S<object>>` which is not considered constructible because it does not implement `IEnumerable`.